### PR TITLE
A few dataframe metadata fixes

### LINKED ISSUE #2564: Index dtype changes when assigning column from index
- Creating a new ddf column from the ddf's existing index erroneously changes the dtype of the index
- Impact: causes downstream ddf operations to fail or behave in unintended ways, e.g. dd.concat

```py
>>> ddf = dd.from_pandas(pd.DataFrame(dict(i=['a','b'], a=[1,2])), npartitions=1)
>>> ddf = ddf.set_index('i)

>>> ddf.index

Dask Index Structure:
npartitions=1
a    object
b       ...
Name: i, dtype: object
Dask Name: sort_index, 4 tasks

>>> ddf.assign(b=ddf.index).index

Dask Index Structure:
npartitions=1
a    int64
b      ...
dtype: int64
Dask Name: assign, 6 tasks
```

## Versions
```sh
$ python --version
Python 3.6.0

$ cat requirements.txt | egrep 'dask/dask' | grep -v '^#'
git+https://github.com/dask/dask.git@00ed56e
```

### PR DESCRIPTION
**Pull request by jcrist**
Due to a bug in `dd.utils.assert_eq`, we weren't properly testing that index metadata was correct. This fixes that, as well as issues this turned up.

- Use `_meta_nonempty` in assign
- Use `_meta_nonempty` in groupby aggregations
- Error nicely when trying to infer metadata from a `dask.delayed`
object (which we can't do since it could be any dtype).
- Fix `dd.utils.assert_eq` to properly compare index dtypes

Fixes #2564.

### CODE REVIEW

**Comment by mrocklin** [comment]
Thanks for finding and resolving this @jcrist !

I'm curious, are the parquet errors related, or are they due to something else?

**Comment by jcrist** [comment]
Not sure, I can take a look tomorrow.

**Comment by jcrist** [comment]
This is due to a bug in our index metadata handling for pyarrow 0.13.0. I spent 20 mins or so trying to debug and haven't found an easy solution (the test fixes here turned up a few bad code paths). With our desire to rewrite all the parquet handling anyway, and other pyarrow 0.13.0 issues (no pickle support) would it be fine to just skip these tests?

**Comment by jcrist** [comment]
Alternatively we could just error nicely if pyarrow 0.13 is installed and recommend people downgrade/upgrade? I'm not sure what the complete parquet & pyarrow 0.13.0 status is, but until we do a clean rewrite there may be too many bugs for me to completely trust the index handling logic here.

**Comment by martindurant** [comment]
Some things are still broken after https://github.com/dask/dask/pull/4668 ?

The general problem with inference from parquet is that there are genuinely many possibilities in the schema contents, the file organisations and further metadata present in the files, and the engines' view of it (the latter of which may depend on version, as in the present case of arrow0.13). So *partly* the code is convoluted because it grew organically to accommodate a moving target, but also in no small part it reflects a genuinely complex situation. I've tried to be as clear as I can about the situation in comments in https://github.com/dask/dask/pull/4336 .
A "clear rewrite" should probably include stricter version requirements. 

Since arrow is actively working on the parquet story for the next release, including pickle and metadata-file handling, I would be OK skipping some tests for now. Whether the rewrite will succeed in making a more maintainable situation remains to be seen.

**Comment by mrocklin** [comment]
> A "clear rewrite" should probably include stricter version requirements.

+1

> So partly the code is convoluted because it grew organically to accommodate a moving target

I recall early conversations when you were writing the first integration where there were open questions of "should this logic go in fastparquet, or in dask.dataframe?" to which we really didn't have any good answers :)

> but also in no small part it reflects a genuinely complex situation

Indeed

> Whether the rewrite will succeed in making a more maintainable situation remains to be seen.

I think that we're more than capable of making this happen if we act as a group.  I don't think that I am independently capable of this though.

I think that the parquet situation is important for Dask.

**Comment by martindurant** [comment]
>> Whether the rewrite will succeed in making a more maintainable situation remains to be seen.

> I think that we're more than capable of making this happen if we act as a group.

I should clarify that by "success", here, I mean greatly reducing the code complexity and maintenance burden. I'm sure we can make something that work well for the current situation :)

**Comment by jcrist** [comment]
> Some things are still broken after #4668 ?

This PR fixes a bug in our test suite that let metadata issues slip through. Yes things are still broken after #4668, and the fix seems to require refactoring the pyarrow code path significantly. To be clear, I'm not attacking the state of the code, I'm just saying that for now it seems easier to either say "don't use pyarrow 0.13.0 with dask" or skip these tests. I don't care which one we choose.

**Comment by mrocklin** [comment]
I'm totally ok with a "don't use pyarrow 0.13.0" with Dask warning

**Comment by martindurant** [comment]
> don't use pyarrow 0.13.0

Agree, since it doesn't work with distributed anyway

**Comment by jcrist** [comment]
Cool, I'll push that here.

**Comment by jcrist** [comment]
I think this is good to go now.

File: dask/dataframe/io/parquet.py [hunk 1]
```diff
@@ -1059,8 +1059,12 @@ def get_engine(engine):
     elif engine == 'pyarrow':
         pa = import_required('pyarrow', "`pyarrow` not installed")
 
-        if LooseVersion(pa.__version__) < '0.8.0':
+        pa_version = LooseVersion(pa.__version__)
+        if pa_version < '0.8.0':
             raise RuntimeError("PyArrow version >= 0.8.0 required")
+        elif pa_version == '0.13.0':
+            raise RuntimeError("PyArrow versino 0.13.0 isn't supported, please "
```
Comment (martindurant) [inline dask/dataframe/io/parquet.py]: spelling

**Comment by jcrist** [comment]
Merging later today if no comment.

**Pull request merged**